Build failure: fpc on x86_64-darwin (upstream incompatibility with Clang 18+)

### Nixpkgs version

- Stable (25.05)

### Steps to reproduce

On an Intel-based Mac:
```console
$ nix-build '<nixpkgs>' -A fpc
```
or using the flake:
```console
$ nix build 'nixpkgs#fpc'
```

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/297287852

### Relevant log output

```console
[…]
Compiling ./rtl-objpas/src/inc/rtti.pp
Writing Resource String Table file: rtti.rsj
Assembling rtti
Warning: supplying the --target x86_64-apple-macosx11.3 != x86_64-apple-darwin argument to a nix-wrapped compiler may not work correctly - cc-wrapper is currently not designed with multi-target compilers in mind. You may want to use an un-wrapped compiler instead.
rtti.pp(4361) Error: Error while assembling exitcode 1
rtti.pp(4361) Fatal: There were 2 errors compiling module, stopping
Fatal: Compilation aborted
[…many, many 'argument unused during compilation' warnings, but they don't affect the rest of the build…]
rtl-objpas/units/x86_64-darwin/rtti.s:12:1: error: non-private labels cannot appear between .cfi_startproc / .cfi_endproc pairs
_$RTTI$_Lj102:
^
rtl-objpas/units/x86_64-darwin/rtti.s:6:1: error: previous .cfi_startproc was here
.cfi_startproc
^

The installer encountered the following error:
Compilation of "BuildUnit_rtl_objpas.pp" failed
make[3]: *** [Makefile:1723: smart] Error 1
make[3]: Leaving directory '/private/tmp/nix-build-fpc-3.2.2.drv-0/fpcbuild-3.2.2/fpcsrc/packages'
make[2]: *** [Makefile:2682: packages_smart] Error 2
make[2]: Leaving directory '/private/tmp/nix-build-fpc-3.2.2.drv-0/fpcbuild-3.2.2/fpcsrc'
make[1]: *** [Makefile:2837: build-stamp.x86_64-darwin] Error 2
make[1]: Leaving directory '/private/tmp/nix-build-fpc-3.2.2.drv-0/fpcbuild-3.2.2/fpcsrc'
make: *** [Makefile:2306: fpcsrc/build-stamp.x86_64-darwin] Error 2
```

### Additional context

I originally mentioned this in <https://github.com/NixOS/nixpkgs/issues/358328#issuecomment-2569576837>, but this is different enough from that issue that I thought it needed to be reported separately.

This seems to be a [known upstream issue](https://gitlab.com/freepascal.org/fpc/source/-/issues/41045) when building using Clang 18+. It started happening here when Nixpkgs changed to using Clang 19 by default (#354107). Switching it back to Clang 17 with `fpc.override { inherit (llvmPackages_17) stdenv; }` seems to let it build correctly for now.

There's an [open merge request](https://gitlab.com/freepascal.org/fpc/source/-/merge_requests/887) on the upstream FPC repo to try and fix this, but it seems to still be a work in progress.

The error seems to happen in x86_64-specific assembly code. Apple Silicon (`aarch64-darwin`) builds are unaffected.

### System metadata

 - system: `"x86_64-darwin"`
 - host os: `Darwin 24.5.0, macOS 15.5`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.28.3`
 - channels(root): `"nixpkgs-25.05-darwin"`
 - nixpkgs: `/nix/store/44im6g6avj288m5wqbnkajhycqbi36f6-nixpkgs-25.05-darwin/nixpkgs`
     - **Note**: Also tested against the current `nixpkgs-unstable` using `NIX_PATH='nixpkgs=https://github.com/NixOS/nixpkgs/archive/refs/heads/nixpkgs-unstable.tar.gz' nix-build '<nixpkgs>' -A fpc`, with the same result.

### Notify maintainers

`fpc` maintainer: @7c6f434c

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -4990,7 +4990,13 @@ with pkgs;
   flutter327 = flutterPackages.v3_27;
   flutter324 = flutterPackages.v3_24;
 
-  fpc = callPackage ../development/compilers/fpc { };
+  fpc = callPackage ../development/compilers/fpc {
+    # https://github.com/NixOS/nixpkgs/issues/416485
+    # TODO: remove when upstream issue is fixed:
+    # https://gitlab.com/freepascal.org/fpc/source/-/issues/41045
+    stdenv =
+      if stdenv.cc.isClang && stdenv.hostPlatform.isx86_64 then llvmPackages_17.stdenv else stdenv;
+  };
 
   gambit = callPackage ../development/compilers/gambit { };
   gambit-unstable = callPackage ../development/compilers/gambit/unstable.nix { };

